File: peer-review/TTK4145-2025-code-peer-review-files/75e5de3e/driver-go-master/requests/requests.go
package requests

import (
	. "Driver-go/elevator"
)

type DirnBehaviourPair struct {
	Direction Dirn
	Behaviour ElevatorBehaviour
}


//Funksjon som sjekker om det er bestillinger i etasjer over
func RequestsAbove(e Elevator) bool {
	for f := e.Floor + 1; f < NFloors; f++ {
		for btn := 0; btn < NBtns; btn++ {
			if e.Requests[f][btn] == 1 {
				return true
			}
		}
	}
	return false
}


//Funksjon som sjekker om det er bestillinger i etasjer under
func RequestsBelow(e Elevator) bool {
	for f := 0; f < e.Floor; f++ {
		for btn := 0; btn < NBtns; btn++ {
			if e.Requests[f][btn] == 1 {
				return true
			}
		}
	}
	return false
}


//Funksjon som sjekker om det er bestillinger i denne etasjen 
func RequestsHere(e Elevator) bool {
	for btn := 0; btn < NBtns; btn++ {
		if e.Requests[e.Floor][btn] == 1 {
			return true
		}
	}
	return false
}


//Funksjon som bestemmer oppførsel til heisen basert på bevegelsesretning og bestillinger
func RequestsChooseDirection(e Elevator) DirnBehaviourPair {
	switch e.Dirn {
	case Up:
		if RequestsAbove(e) {
			return DirnBehaviourPair{Direction: Up, Behaviour: Moving}
		} else if RequestsHere(e) {
			return DirnBehaviourPair{Direction: Down, Behaviour: DoorOpen}
		} else if RequestsBelow(e) {
			return DirnBehaviourPair{Direction: Down, Behaviour: Moving}
		}
		return DirnBehaviourPair{Direction: Stop, Behaviour: Idle}

	case Down:
		if RequestsBelow(e) {
			return DirnBehaviourPair{Direction: Down, Behaviour: Moving}
		} else if RequestsHere(e) {
			return DirnBehaviourPair{Direction: Up, Behaviour: DoorOpen}
		} else if RequestsAbove(e) {
			return DirnBehaviourPair{Direction: Up, Behaviour: Moving}
		}
		return DirnBehaviourPair{Direction: Stop, Behaviour: Idle}

	case Stop:
		if RequestsHere(e) {
			return DirnBehaviourPair{Direction: Stop, Behaviour: DoorOpen}
		} else if RequestsAbove(e) {
			return DirnBehaviourPair{Direction: Up, Behaviour: Moving}
		} else if RequestsBelow(e) {
			return DirnBehaviourPair{Direction: Down, Behaviour: Moving}
		}
		return DirnBehaviourPair{Direction: Stop, Behaviour: Idle}

	default:
		return DirnBehaviourPair{Direction: Stop, Behaviour: Idle}
	}
}


//Funksjon som sjekker om heisen skal stoppe i nåværende etasje
func RequestsShouldStop(e Elevator) bool {
	switch e.Dirn {
	case Down:
		if e.Requests[e.Floor][HallDown] == 1 ||
			e.Requests[e.Floor][Cab] == 1 ||
			!RequestsBelow(e) {
			return true
		} else {
			return false
		}

	case Up:
		if e.Requests[e.Floor][HallUp] == 1 ||
			e.Requests[e.Floor][Cab] == 1 ||
			!RequestsAbove(e) {
			return true
		} else {
			return false
		}
	case Stop:
		fallthrough
	default:
		return true
	}
}


//Funksjon som sletter en ny bestilling dersom bestillingen er fra nåværende etasje 
func RequestsShouldClearImmediately(e Elevator, btnFloor int) bool {
	return e.Floor == btnFloor
}


//Funksjon for å slette bestillinger som er utført på nåværende etasje
func RequestsClearAtCurrentFloor(e Elevator) Elevator {
	//Sletter cabcalls fra nåværende etasje
	e.Requests[e.Floor][Cab] = 0
	//Switch case for å bestemme hvilke hallcalls som skal slettes
	switch e.Dirn {
	case Up:
		if !RequestsAbove(e) && e.Requests[e.Floor][HallUp] == 0 {
			e.Requests[e.Floor][HallDown] = 0
		}
		e.Requests[e.Floor][HallUp] = 0

	case Down:
		if !RequestsBelow(e) && e.Requests[e.Floor][HallDown] == 0 {
			e.Requests[e.Floor][HallUp] = 0
		}
		e.Requests[e.Floor][HallDown] = 0

	case Stop:
		fallthrough
	default:
		e.Requests[e.Floor][HallUp] = 0
		e.Requests[e.Floor][HallDown] = 0
	}

	return e
}

File: peer-review/TTK4145-2025-code-peer-review-files/b67214ff/single_elevator/src/execution/requests.rs
use crate::execution::elevator::{self, Button, Dirn, Elevator, ElevatorBehaviour, N_BUTTONS, N_FLOORS};


pub struct DirnBehaviourPair {
    pub dirn : Dirn,
    pub behaviour : ElevatorBehaviour
}

pub fn requests_above(e : &Elevator) -> bool {
    for f in (e.floor.expect("Floor is None")+1) as usize..N_FLOORS {
        for btn in 0..N_BUTTONS {
            if e.requests[f][btn] {
                return true;
            }
        }
    }
    return false;
}

pub fn requests_below(e : &Elevator) -> bool {
    for f in 0..e.floor.expect("Floor is None") {
        for btn in 0..N_BUTTONS {
            if e.requests[f as usize][btn] {
                return true;
            }
        }
    }
    return false;
}

pub fn requests_here(e : &Elevator) -> bool {
    for btn in 0..N_BUTTONS {
        if e.requests[e.floor.expect("Floor is None") as usize][btn] {
            return true;
        }
    }
    return false;
}

pub fn requests_choose_direction(e : &Elevator) -> DirnBehaviourPair {
    match e.dirn {
        Dirn::Up => {
            return if requests_above(&e) {DirnBehaviourPair{dirn : Dirn::Up , behaviour : ElevatorBehaviour::Moving}}
                else if requests_here(&e) {DirnBehaviourPair{dirn : Dirn::Down , behaviour : ElevatorBehaviour::DoorOpen}}
                else if requests_below(&e) {DirnBehaviourPair{dirn : Dirn::Down , behaviour : ElevatorBehaviour::Moving}}
                else { DirnBehaviourPair{dirn : Dirn::Stop , behaviour : ElevatorBehaviour::Idle}}
        }
        Dirn::Down => {
            return if requests_below(&e) {DirnBehaviourPair{dirn : Dirn::Down , behaviour : ElevatorBehaviour::Moving}}
                else if requests_here(&e) {DirnBehaviourPair{dirn : Dirn::Up , behaviour : ElevatorBehaviour::DoorOpen}}
                else if requests_above(&e) {DirnBehaviourPair{dirn : Dirn::Up , behaviour : ElevatorBehaviour::Moving}}
                else { DirnBehaviourPair{dirn : Dirn::Stop , behaviour : ElevatorBehaviour::Idle}}

        }
        Dirn::Stop => {
            return if requests_here(&e) {DirnBehaviourPair{dirn : Dirn::Stop , behaviour : ElevatorBehaviour::DoorOpen}}
                else if requests_above(&e) {DirnBehaviourPair{dirn : Dirn::Up , behaviour : ElevatorBehaviour::Moving}}
                else if requests_below(&e) {DirnBehaviourPair{dirn : Dirn::Down , behaviour : ElevatorBehaviour::Moving}}
                else { DirnBehaviourPair{dirn : Dirn::Stop , behaviour : ElevatorBehaviour::Idle}}

        }
        _ => {
            eprintln!("Request_choose_direction function in request.rs got invalid direction from elevator");
            return DirnBehaviourPair{dirn : Dirn::Stop , behaviour : ElevatorBehaviour::Idle}
        }
    }
}


pub fn requests_should_stop(e : &Elevator) -> bool {
    match e.dirn {
        Dirn::Down => {
            return 
                e.requests[e.floor.expect("Floor is None") as usize][Button::HallDown as usize] ||
                e.requests[e.floor.expect("Floor is None") as usize][Button::Cab as usize] ||
                !requests_below(&e);
        }
        Dirn::Up => {
            return 
                e.requests[e.floor.expect("Floor is None")as usize][Button::HallUp as usize] ||
                e.requests[e.floor.expect("Floor is None")as usize][Button::Cab as usize]  ||
                !requests_above(&e);

        }
        Dirn::Stop => {
            return true;
        }
        _ => {
            eprintln!("Request_should_stop function in request.rs got invalid direction from elevator");
            return true;
        }

    }
}


pub fn requests_should_clear_immediately_matrix(e : &mut Elevator ) -> bool { //Not great that this now changes the elevator requests. Might be a problem
    let mut should_clear = false;

    for f in 0..N_FLOORS {
        for btn_type in  0..N_BUTTONS {
            if (e.floor.expect("Floor is None") == f as u8) && (
                (e.dirn == Dirn::Up && btn_type == Button::HallUp as usize) ||
                (e.dirn == Dirn::Down && btn_type == Button::HallDown as usize) ||
                e.dirn == Dirn::Stop ||
                btn_type == Button::Cab as usize
            ) {
                should_clear = true;
                e.requests[f][btn_type] = false;
            }
        }  
    }
    return should_clear
} 




pub fn requests_clear_at_current_floor(elevator: &mut elevator::Elevator) {
    
    
    let floor = elevator.floor.expect("Floor is None") as usize;
    let dirn = elevator.dirn;
    let requests_above_exist = requests_above(elevator);
    let requests_below_exist = requests_below(elevator);
    
    elevator.requests[floor as usize][Button::Cab as usize] = false; //Should this be sent to the controller?
    
    match dirn {
        Dirn::Up => {
            if !requests_above_exist && !elevator.requests[floor][Button::HallUp as usize] {
                elevator.requests[floor][Button::HallDown as usize] = false;
            }
            elevator.requests[floor][Button::HallUp as usize] = false;
        }
        Dirn::Down => {
            if !requests_below_exist && !elevator.requests[floor][Button::HallDown as usize] {
                elevator.requests[floor][Button::HallUp as usize] = false;
            }
            elevator.requests[floor][Button::HallDown as usize] = false;
            
        }
        Dirn::Stop => {
            elevator.requests[floor][Button::HallUp as usize] = false;
            elevator.requests[floor][Button::HallDown as usize] = false;
        }
    }
}
